text/0806-engine-router-helpers.md

```diff
@@ -0,0 +1,273 @@
+2. Replace use of `<LinkTo @route="inner-route" />` component with `<LinkTo @route={{path-for "inner-route"}} />` in every engine codebase.
+3. Replace use of `<LinkToExternal @route="external-route" />` component with `<LinkTo @route={{path-for "external-route"}} />` in every engine codebase.
+4. If there are any addons providing some shared code that is being used by engine, steps #2 and #3 needs to be applied to addons as well.
+
+<!--- 
+> What names and terminology work best for these concepts and why? How is this
+idea best presented? As a continuation of existing Ember patterns, or as a
+wholly new one?
+
+> Would the acceptance of this proposal mean the Ember guides must be
+re-organized or altered? Does it change how Ember is taught to new users
+at any level?
+
+> How should this feature be introduced and taught to the existing Ember users?
+-->
+
+## Drawbacks
+
+> Why should we *not* do this? Please consider the impact on teaching Ember,
+on the integration of this feature with other existing and planned features,
+on the impact of the API churn on existing apps, etc.
+
+> There are tradeoffs to choosing any path, please attempt to identify them here.
+
+Changes proposed in this RFC span multiple existing packages and require users to install extra package.
+Such churn deemed to be minor and codemode can be provided.
+
+Users of Ember Engine would need to use more verbose syntax to achieve linking between routes of the same engine:
+
+```handlebars
+{{!-- Before. --}}
+<LinkTo @route="contact">Contact Us</LinkTo>
+
+{{!-- After. --}}
+<LinkTo @route={{path-for "contact"}}>Contact Us</LinkTo>
+```
+
+This deemed to be minor inconvenience for the code bases using Ember Engines.
+
+## Alternatives
+
+* Utilize `@ember/legacy-built-in-components` to keep `<LinkToExternal />` a thin wrapper around legacy implementation of `<LinkTo />` component.
+
+  As this may be considered temporary solution, this is not viable long term solution as any changes made to `<LinkTo />` component in the Ember.js codebase will need to be backported to `@ember/legacy-built-in-components` which is not planned to be done.
+
+* Re-implement `<LinkTo />` functionality in `ember-engine`.
+
+  This is not viable long term solution as any changes made to `<LinkTo />` component in the Ember.js codebase would need to be backported to `@ember/legacy-built-in-components` which is not planned to be done.
+
+* Deprecate and remove `externalRoutes` mapping.
+
+  This would require providing *full route name* to `<LinkTo />` component *including* engine mount point. This option is not viable as it breaks isolation principle of Ember Engines whereas every engine should be self-scoped micro-frontend.
+
+* Bake in these implementation details into Ember.js codebase.
+
+  Concerns discussed in this RFC are specific to Ember Engines and there is no need to ship this functionality to every user of Ember.js.
```

Comment:
So basically what `ember-engines-router-service` does plus wrapping `<LinkTo />` to use said service?

Reply:
Somewhat, but not quite. [`ember-engines-router-service`](https://github.com/villander/ember-engines-router-service#readme) does not unify the internal and external route namespaces to drop `*External` methods. Therefore it is not interoperable with the standard `RouterService`, as you still have to invoke e.g. `transitionToExternal(externalRouteName)` in order to navigate to an external route.

The two linked issues contain a discussion regarding the pros and cons of this namespace merger.

If `ember-engines-router-service` were to merge the namespaces, such that you can use `transitionTo(externalRouteName)` to navigate to an external route, then yes: The only remaning change is refactoring / wrapping / polyfilling `<LinkTo>` to only use the `RouterService` public API.

I think that refactoring `<LinkTo>` to only rely on public `RouterService` API is a sensible change either way. And if `ember-engines` ships its own `EngineRouterService` as `service:router`, then there's no need for additional dependencies and no need for third-party authors to worry about engine compatibility, because it will work out of the box.